### PR TITLE
Merge the ES6 feature branch back into master

### PR DESCRIPTION
**Pull request by MrTimscampi**
**Changes**

Pretty self-explanatory.

Only a review of d822366 should be needed.

**Issues**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -107,38 +143,128 @@
...
+          "src/elements/emby-progressbar/emby-progressbar.js",
+          "src/elements/emby-progressring/emby-progressring.js",
+          "src/elements/emby-radio/emby-radio.js",
+          "src/elements/emby-radio/emby-radio.js",
+          "src/elements/emby-ratingbutton/emby-ratingbutton.js",
+          "src/elements/emby-scrollbuttons/emby-scrollbuttons.js",
+          "src/elements/emby-scroller/emby-scroller.js",
+          "src/elements/emby-scroller/emby-scroller.js",
+          "src/elements/emby-select/emby-select.js",
+          "src/elements/emby-slider/emby-slider.js",
```
[the remark this answers is not in the document]
Comment (dmitrylyzo) [reply package.json]: Why remove `emby-slider.js`?

**Comment by dmitrylyzo** [comment]
https://github.com/jellyfin/jellyfin-web/blob/ffc22f629e31d7739b8ebef0f8db471b23b83f1d/src/components/alphaPicker/alphaPicker.js#L216
https://github.com/jellyfin/jellyfin-web/blob/ffc22f629e31d7739b8ebef0f8db471b23b83f1d/src/components/alphaPicker/alphaPicker.js#L227
`onAlphaPickerClick.bind(this) !== onAlphaPickerClick.bind(this)`
So I suppose that won't work as expected. But looks like it is here from the ages and isn't used.

@MrTimscampi return `emby-slider.js` into `package.json`.

File: src/controllers/user/display.js [hunk 2]
```diff
@@ -24,7 +24,7 @@ import autoFocuser from 'autoFocuser';
             if (settingsInstance) {
                 settingsInstance.loadData();
             } else {
-                settingsInstance = new DisplaySettings({
+                settingsInstance = new DisplaySettings.default({
```
Comment (dmitrylyzo) [inline src/controllers/user/display.js:27]:
```suggestion
                settingsInstance = new DisplaySettings({
```
I got `_displaySettings.default.default`. :man_shrugging:

File: src/components/search/searchfields.js [hunk 3]
```diff
@@ -43,7 +53,7 @@ define(['layoutManager', 'globalize', 'require', 'events', 'browser', 'alphaPick
 
     function initAlphaPicker(alphaPickerElement, instance) {
 
-        instance.alphaPicker = new AlphaPicker({
+        instance.alphaPicker = new AlphaPicker.default({
```
Comment (dmitrylyzo) [inline src/components/search/searchfields.js:56]:
```suggestion
        instance.alphaPicker = new AlphaPicker({
```
Looks like `.default` is not required to import ES6 into ES6.

File: src/components/alphaPicker/alphaPicker.js [hunk 4]
```diff
@@ -95,227 +106,228 @@ define(['focusManager', 'layoutManager', 'dom', 'css!./style.css', 'paper-icon-b
...
 
-            alphaFocusTimeout = null;
+                alphaFocusTimeout = null;
 
-            if (document.activeElement === alphaFocusedElement) {
-                var value = alphaFocusedElement.getAttribute('data-value');
-                self.value(value, true);
+                if (document.activeElement === alphaFocusedElement) {
+                    const value = alphaFocusedElement.getAttribute('data-value');
+                    this.value(value, true);
```
Comment (dmitrylyzo) [inline src/components/alphaPicker/alphaPicker.js:135]:
```
Uncaught TypeError: this.value is not a function
    at onAlphaFocusTimeout (alphaPicker.js:135)
```
`this` is `Window`. Occurred on TV layout only.

Comment (MrTimscampi) [reply src/components/alphaPicker/alphaPicker.js:135]: Hm, could it be because it's a callback from a timeout?

File: src/scripts/playlists.js [hunk 5]
```diff
@@ -1,195 +1,202 @@
...
+    }
+
+    const data = {};
+    view.addEventListener('viewbeforeshow', function () {
+        reloadItems();
+    });
+    view.querySelector('.btnNewPlaylist').addEventListener('click', function () {
+        import('playlistEditor').then(({default: playlistEditor}) => {
+            const serverId = ApiClient.serverInfo().Id;
+            new playlistEditor.showEditor({
```
Comment (dmitrylyzo) [inline src/scripts/playlists.js:194]: `showEditor` is `default`
```suggestion
            new playlistEditor({
```
But I cannot test this - it seems that this file isn't used.

File: package.json [hunk 6]
```diff
@@ -91,13 +91,49 @@
...
+          "src/components/confirm/confirm.js",
+          "src/components/dialog/dialog.js",
+          "src/components/dialogHelper/dialogHelper.js",
+          "src/components/directorybrowser/directorybrowser.js",
+          "src/components/directorybrowser/directorybrowser.js",
+          "src/components/displaySettings/displaySettings.js",
+          "src/components/fetchhelper.js",
+          "src/components/fetchhelper.js",
+          "src/components/filterdialog/filterdialog.js",
+          "src/components/filterdialog/filterdialog.js",
```
[the remark this answers is not in the document]
Comment (dmitrylyzo) [reply package.json:112]: I unresolved this because it is good suggestion.

**Comment by dmitrylyzo** [comment]
```diff
--- a/src/controllers/edititemmetadata.js
+++ b/src/controllers/edititemmetadata.js
@@ -5,7 +5,7 @@ define(['loading', 'scripts/editorsidebar'], function (loading) {
         loading.show();
 
         if (itemId) {
-            require(['metadataEditor'], function (metadataEditor) {
+            require(['metadataEditor'], function ({default: metadataEditor}) {
                 metadataEditor.embed(context.querySelector('.editPageInnerContent'), itemId, ApiClient.serverInfo().Id);
             });
         } else {
```

Only alphaPicker and another runtime error remain.

File: src/controllers/dashboard/dlna/profile.js [hunk 7]
```diff
@@ -244,12 +252,11 @@ define(['jQuery', 'loading', 'globalize', 'emby-select', 'emby-button', 'emby-in
...
     function renderDirectPlayProfiles(page, profiles) {
-        var html = '';
+        let html = '';
         html += '<ul data-role="listview" data-inset="true" data-split-icon="delete">';
-        var currentType;
+        let currentType;
 
-        for (var i = 0, length = profiles.length; i < length; i++) {
-            var profile = profiles[i];
+        for (const profile of profiles) {
```
[the remark this answers is not in the document]
Comment (Camc314) [reply src/controllers/dashboard/dlna/profile.js:259]: I think this is fixed now?

File: src/controllers/dashboard/users/useredit.js [hunk 8]
```diff
@@ -1,26 +1,26 @@
...
-            var i;
-            var length;
-            var folder;
-            var isChecked;
-            var checkedAttribute;
-            var html = '';
-
-            for (i = 0, length = mediaFolders.length; i < length; i++) {
-                folder = mediaFolders[i];
+            let folder;
```
Comment (Camc314) [inline src/controllers/dashboard/users/useredit.js:12]:
```suggestion
```

File: src/controllers/dashboard/dlna/profile.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (dmitrylyzo) [reply src/controllers/dashboard/dlna/profile.js:259]: Fails here
https://github.com/jellyfin/jellyfin-web/blob/5799a877a9ea909f8fadc51bbe734551e6dc5eae/src/controllers/dashboard/dlna/profile.js#L266-L268

File: src/components/metadataEditor/metadataEditor.js [hunk 9]
```diff
@@ -212,19 +233,19 @@ define(['itemHelper', 'dom', 'layoutManager', 'dialogHelper', 'datetime', 'loadi
     function addElementToList(source, sortCallback) {
         require(['prompt'], function (prompt) {
```
Comment (Camc314) [inline src/components/metadataEditor/metadataEditor.js:234]:
```suggestion
        import('prompt').then(({default: prompt}) => {
```

File: src/components/metadataEditor/metadataEditor.js [hunk 10]
```diff
@@ -1058,7 +1077,7 @@ define(['itemHelper', 'dom', 'layoutManager', 'dialogHelper', 'datetime', 'loadi
 
     function centerFocus(elem, horiz, on) {
         require(['scrollHelper'], function (scrollHelper) {
```
Comment (Camc314) [inline src/components/metadataEditor/metadataEditor.js:1079]:
```suggestion
        import('scrollHelper').then(({default: scrollHelper}) => {
```

File: src/components/metadataEditor/metadataEditor.js [hunk 11]
```diff
@@ -1068,7 +1087,7 @@ define(['itemHelper', 'dom', 'layoutManager', 'dialogHelper', 'datetime', 'loadi
 
         require(['text!./metadataEditor.template.html'], function (template) {
```
Comment (Camc314) [inline src/components/metadataEditor/metadataEditor.js:1088]:
```suggestion 
        import('text!./metadataEditor.template.html').then(({default: template}) => {
```

File: src/components/metadataEditor/metadataEditor.js [hunk 12]
```diff
@@ -1125,7 +1144,7 @@ define(['itemHelper', 'dom', 'layoutManager', 'dialogHelper', 'datetime', 'loadi
 
                 require(['text!./metadataEditor.template.html'], function (template) {
```
Comment (Camc314) [inline src/components/metadataEditor/metadataEditor.js:1145]:
```suggestion
                import('text!./metadataEditor.template.html').then(({default: template}) => {
```

File: src/components/metadataEditor/metadataEditor.js (hunk 9 above)
Comment (dmitrylyzo) [reply src/components/metadataEditor/metadataEditor.js:234]: This requires to change `prompt.default` down below.

File: src/components/metadataEditor/metadataEditor.js [hunk 13]
```diff
@@ -212,19 +233,19 @@ define(['itemHelper', 'dom', 'layoutManager', 'dialogHelper', 'datetime', 'loadi
     function addElementToList(source, sortCallback) {
         require(['prompt'], function (prompt) {
 
-            prompt({
+            prompt.default({
```
Comment (Camc314) [inline src/components/metadataEditor/metadataEditor.js:236]:
```suggestion
            prompt({

File: src/controllers/dashboard/dlna/profile.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (MrTimscampi) [reply src/controllers/dashboard/dlna/profile.js:259]: Usually for these, you need something like this:

`for (const [index, profile] of profiles.entries()) {`

I just didn't see that the index was referenced :)

**Comment by MrTimscampi** [comment]
> `focus` is called before searchfield appears in HTML.

@dmitrylyzo Is this really blocking ? It doesn't seem directly related to the ES6 branch, since `searchtab.js` is not yet migrated.

I'd like to get this merged asap, so we can continue merging the backlog.

Same for all the `for`, we already deal with all of these in #1460 so I'd rather take care of them there than hold up this PR.

**Comment by anthonylavado** [comment]
Let there be ES6.

**Pull request merged**